### PR TITLE
network: support ID_NET_MANAGED_BY udev property

### LINKED ISSUE #29768: RFE: networkd: honour a new ID_NET_MANAGED_BY udev propery or so
Right now NetworkManager and networkd are happy to fight for ownership of network interfaces. That kinda sucks.

Here's an idea: let's make both unconditionally honour a NET_MANAGED_BY= udev property, and if set to any string that is different from some identifier for the respective project, the interfaces will be left unmanaged.

And example:

1. systemd-networkd would leave all interfaces unmanaged where NET_MANAGED_BY= is set to a string other than io.systemd.networkd
2. NetworkManager would leave all interfaces unmanaged where NET_MANAGED_BY= is set set to a string other than dev.networkmanager.nm

If the property is not set we'll live in the wild-west as before. 

What's the usecase?

Allowing the admin to make an explicit choice which project shall manage which interfaces. (Could even be generated dynamically via rules files in /run/)

What's not the usecase?

This is not a locking concept: if both NM and networkd are told to configure a device then they will still fight for it.

### PR DESCRIPTION
**Pull request by yuwata**
Closes #29768.

### CODE REVIEW

File: src/network/networkd-link.c [hunk 1]
```diff
@@ -1499,6 +1500,15 @@ int manager_udev_process_link(Manager *m, sd_device *device, sd_device_action_t
                 return 0;
         }
 
+        r = sd_device_get_property_value(device, "NET_MANAGED_BY", &s);
+        if (r < 0 && r != -ENOENT)
+                log_device_debug_errno(device, r, "Failed to get NET_MANAGED_BY udev property, ignoring: %m");
+        if (r >= 0 && !streq(s, "io.systemd.Network")) {
+                log_device_debug(device, "Interface is requested to be managed by '%s', unmanaging the interface.", s);
```
Comment (poettering) [inline src/network/networkd-link.c]: unmanaging → not managing

**Comment by poettering** [comment]
Looks excellent! thanks for working on this.

we probably should document this somewhere. Maybe as a new paragraph in the `[MATCH] Section Options` section of systemd.network(5). Maybe something like this:

```
<para>Note that any network interface that has the <varname>NET_MANAGED_BY=</varname> udev property set will never be matched by any .network file – unless the property's value is the string <literal>io.systemd.Network</literal> – even if the <literal>[Match]</literal> section would otherwise match. This may be used to exclude specific network interfaces from networkd's management, while keeping the <literal>[Match]</literal> section generic. The `<varname>NET_MANAGED_BY=</varname> thus declares intended <emphasis>ownership</emphasis> of the device, and permits ensuring that concurrent network management implementations do not compete for management of specific devices.</para>
````
or something like that.

**Comment by poettering** [comment]
hmm, maybe we should stick to our usual rule of prefixing props with `ID_` for this actually, hence, ID_NET_MANAGED_BY=?

just to be on the safe side, to leave the unprefix properties to the kernel.

**Comment by yuwata** [comment]
@poettering Thank you for the suggestion about the document. Renamed to ID_NET_MANAGED_BY=. Upgrading the green label.

**Comment by poettering** [comment]
@thom311 the ball is now in your court! ;-)

**Comment by thom311** [comment]
This will be supported by future NetworkManager 1.46 release too.

**Pull request merged**